src/main/java/com/google/firebase/internal/RetryConfig.java

```diff
@@ -0,0 +1,160 @@
+/*
+ * Copyright 2019 Google Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.google.firebase.internal;
+
+import static com.google.common.base.Preconditions.checkArgument;
+import static com.google.common.base.Preconditions.checkNotNull;
+
+import com.google.api.client.util.BackOff;
+import com.google.api.client.util.ExponentialBackOff;
+import com.google.api.client.util.Sleeper;
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.collect.ImmutableList;
+import java.util.List;
+import java.util.concurrent.TimeUnit;
+
+/**
+ * Configures when and how HTTP requests should be retried.
+ */
+public final class RetryConfig {
```

Comment:
Did you take a look at just using https://github.com/googleapis/gax-java/blob/master/gax/src/main/java/com/google/api/gax/retrying/RetrySettings.java ?

I think this would be a huge win (if feasible) and we are already using it in Firestore.

Reply:
As far as I can understand gax is tied to GRPC, whereas our HTTP support is provided by the Google API client (everywhere except Firestore). Therefore we have to implement retries using the interfaces provided by the Google API client. We might be able to implement those interfaces using gax algorithm implementations, but I'd rather do this using just one dependency instead of two.

As for `RetrySettings`, it certainly does have all the parameters we'd like to expose. But I'd rather shield the individual classes in the Admin SDK from the dependencies (both gax and API client) by implementing our own configuration class. The way I've implemented this now, individual classes like `FirebaseAuth` and `FirebaseMessaging` don't have to know how the retries are implemented, and they don't have to take a hard dependency on API client. They just pass a `RetryConfig` (which is an interface we control), and the `internal` package provides retries in a single place.

In the future, if we decide to change how retries are implemented (say we manage to migrate to gax), we just need to change the `internal` package.